File: mobile/lib/features/accounting/statement_pdf.dart
import 'package:flutter/services.dart' show rootBundle;
import 'package:pdf/pdf.dart';
import 'package:pdf/widgets.dart' as pw;
import 'package:printing/printing.dart';

import 'account_statement.dart';

/// CARİ HESAP EKSTRESİ — yazdırılabilir belge (web `lib/statementPdf.ts` paritesi).
///
/// Ekrandaki belge ile BİREBİR aynı düzen: kurum başlığı, cari bilgi ızgarası,
/// Tarih / İşlem Türü / Borç / Alacak / Bakiye tablosu, toplam + bakiye bandı,
/// tutarın yazıyla okunuşu. Rakamlar `buildAccountStatement`ten hazır gelir — bu dosya
/// HESAP YAPMAZ, yalnız dizer (iki yerde hesap yapılsaydı ekran ile kâğıt ayrışabilirdi).
///
/// Font: KVKK/onam belgeleriyle AYNI bundle font (Carlito) — internetsiz cihazda da Türkçe
/// karakterler doğru render eder.
class StatementPdf {
  static const _ink = PdfColor.fromInt(0xFF241C21);
  static const _soft = PdfColor.fromInt(0xFF5D4C55);
  static const _muted = PdfColor.fromInt(0xFF7A6873);
  static const _line = PdfColor.fromInt(0xFFDED5DA);
  static const _band = PdfColor.fromInt(0xFFF4EFF1);
  static const _zebra = PdfColor.fromInt(0xFFFBF8F9);
  static const _plum = PdfColor.fromInt(0xFFA5556E);
  static const _debt = PdfColor.fromInt(0xFF9F1239);
  static const _credit = PdfColor.fromInt(0xFF15694A);

  static Future<void> share({
    required AccountStatement doc,
    required String institutionName,
    String? institutionPhone,
    String? institutionEmail,
    String? branchName,
    required String customerCode,
    required String customerName,
    String? customerPhone,
    required int saleCount,
    required String periodLabel,
  }) async {
    final regular = pw.Font.ttf(await rootBundle.load('assets/fonts/KvkkSans.ttf'));
    final bold = pw.Font.ttf(await rootBundle.load('assets/fonts/KvkkSans-Bold.ttf'));
    final theme = pw.ThemeData.withFont(base: regular, bold: bold);
    final issuedAt = formatDocDateTime(DateTime.now());
    final closingDebt = doc.closing >= 0;

    pw.Widget infoRow(String label, String value) => pw.Padding(
          padding: const pw.EdgeInsets.symmetric(vertical: 1.5),
          child: pw.Row(crossAxisAlignment: pw.CrossAxisAlignment.start, children: [
            pw.SizedBox(
              width: 74,
              child: pw.Text(label, style: const pw.TextStyle(fontSize: 8.5, color: _muted)),
            ),
            pw.Text(': ', style: const pw.TextStyle(fontSize: 8.5, color: _muted)),
            pw.Expanded(
              child: pw.Text(value.isEmpty ? '—' : value,
                  style: pw.TextStyle(fontSize: 9, fontWeight: pw.FontWeight.bold, color: _ink)),
            ),
          ]),
        );

    pw.Widget cell(String text,
            {bool right = false, bool boldText = false, PdfColor color = _soft, double size = 8.5}) =>
        pw.Padding(
          padding: const pw.EdgeInsets.symmetric(horizontal: 5, vertical: 4),
          child: pw.Text(
            text,
            textAlign: right ? pw.TextAlign.right : pw.TextAlign.left,
            style: pw.TextStyle(
              fontSize: size,
              color: color,
              fontWeight: boldText ? pw.FontWeight.bold : pw.FontWeight.normal,
            ),
          ),
        );

    // BEŞ SÜTUN — "İşlem Türü" ile "Açıklama" tek sütunda birleşti (`row.label`).
    // Aşağıdaki her TableRow bu sayıya göre elle kuruluyor: sütun sayısı değişirse boş hücre
    // adetleri de değişmeli, yoksa tablo sessizce kayar.
    const widths = <int, pw.TableColumnWidth>{
      0: pw.FixedColumnWidth(52),
      1: pw.FlexColumnWidth(),
      2: pw.FixedColumnWidth(62),
      3: pw.FixedColumnWidth(62),
      4: pw.FixedColumnWidth(68),
    };

    final tableRows = <pw.TableRow>[
      pw.TableRow(
        decoration: const pw.BoxDecoration(
          color: _band,
          border: pw.Border(bottom: pw.BorderSide(color: _plum, width: 1.2)),
        ),
        children: [
          cell('Tarih', boldText: true, color: _ink),
          cell('İşlem Türü', boldText: true, color: _ink),
          cell('Borç (TL)', right: true, boldText: true, color: _ink),
          cell('Alacak (TL)', right: true, boldText: true, color: _ink),
          cell('Bakiye (TL)', right: true, boldText: true, color: _ink),
        ],
      ),
    ];

    if (doc.rows.isEmpty) {
      tableRows.add(pw.TableRow(children: [
        cell(''),
        cell('Bu dönemde hareket bulunmuyor.', color: _muted),
        cell(''),
        cell(''),
        cell(''),
      ]));
    }

    for (var i = 0; i < doc.rows.length; i++) {
      final row = doc.rows[i];
      tableRows.add(pw.TableRow(
        decoration: pw.BoxDecoration(
          color: i.isOdd ? _zebra : null,
          border: const pw.Border(bottom: pw.BorderSide(color: _line, width: .4)),
        ),
        children: [
          cell(formatDocDate(row.date)),
          cell(row.label),
          cell(formatStatementAmount(row.debit), right: true, color: _ink),
          cell(formatStatementAmount(row.credit), right: true, color: _ink),
          cell(formatStatementAmount(row.balance), right: true, boldText: true, color: _ink),
        ],
      ));
    }

    // TOPLAM: yalnız borç/alacak — bakiye zaten son satırda yazılıdır.
    tableRows.add(pw.TableRow(children: [
      cell(''),
      cell('Toplam', right: true, boldText: true, color: _ink, size: 9),
      cell(formatStatementAmount(doc.totalDebit), right: true, boldText: true, color: _ink, size: 9),
      cell(formatStatementAmount(doc.totalCredit), right: true, boldText: true, color: _ink, size: 9),
      cell(''),
    ]));
    tableRows.add(pw.TableRow(children: [
      cell(''),
      cell(''),
      cell(''),
      cell('Bakiye', right: true, boldText: true, color: _ink, size: 9),
      cell('${formatStatementAmount(doc.closing.abs())} TL',
          right: true, boldText: true, color: closingDebt ? _debt : _credit, size: 11),
    ]));

    final pdfDoc = pw.Document(
      title: 'Cari Hesap Ekstresi — $customerName',
      author: institutionName,
      creator: 'BeautyAsist',
    );

    pdfDoc.addPage(pw.MultiPage(
      theme: theme,
      pageFormat: PdfPageFormat.a4,
      margin: const pw.EdgeInsets.fromLTRB(32, 32, 32, 36),
      footer: (context) => pw.Container(
        margin: const pw.EdgeInsets.only(top: 12),
        child: pw.Row(mainAxisAlignment: pw.MainAxisAlignment.spaceBetween, children: [
          pw.Text('$institutionName • Cari Hesap Ekstresi • $customerCode',
              style: const pw.TextStyle(fontSize: 7.5, color: _muted)),
          pw.Text('Sayfa ${context.pageNumber} / ${context.pagesCount}',
              style: const pw.TextStyle(fontSize: 7.5, color: _muted)),
        ]),
      ),
      build: (context) => [
        // ---------- KURUM BAŞLIĞI ----------
        pw.Row(
          crossAxisAlignment: pw.CrossAxisAlignment.start,
          mainAxisAlignment: pw.MainAxisAlignment.spaceBetween,
          children: [
            pw.Expanded(
              child: pw.Column(crossAxisAlignment: pw.CrossAxisAlignment.start, children: [
                pw.Text(institutionName,
                    style: pw.TextStyle(fontSize: 15, fontWeight: pw.FontWeight.bold, color: _ink)),
                if ((branchName ?? '').isNotEmpty)
                  pw.Text(branchName!, style: const pw.TextStyle(fontSize: 8.5, color: _muted)),
              ]),
            ),
            pw.Column(crossAxisAlignment: pw.CrossAxisAlignment.end, children: [
              pw.Text(institutionName.toUpperCase(),
                  style: pw.TextStyle(fontSize: 10, fontWeight: pw.FontWeight.bold, color: _ink)),
              if ((institutionPhone ?? '').isNotEmpty)
                pw.Text('Tel: $institutionPhone', style: const pw.TextStyle(fontSize: 8, color: _muted)),
              if ((institutionEmail ?? '').isNotEmpty)
                pw.Text(institutionEmail!, style: const pw.TextStyle(fontSize: 8, color: _muted)),
            ]),
          ],
        ),
        pw.SizedBox(height: 6),
        pw.Container(height: 1.6, color: _plum),

        // ---------- BAŞLIK ----------
        pw.SizedBox(height: 14),
        pw.Center(
          child: pw.Text('CARİ HESAP EKSTRESİ',
              style: pw.TextStyle(fontSize: 15, fontWeight: pw.FontWeight.bold, color: _ink)),
        ),
        pw.SizedBox(height: 14),

        // ---------- CARİ BİLGİLERİ ----------
        pw.Row(crossAxisAlignment: pw.CrossAxisAlignment.start, children: [
          pw.Expanded(
            child: pw.Column(crossAxisAlignment: pw.CrossAxisAlignment.start, children: [
              infoRow('Cari Kodu', customerCode),
              infoRow('Adı Soyadı', customerName),
              infoRow('Telefon', customerPhone ?? ''),
              infoRow('Kayıtlı Satış', '$saleCount satış'),
            ]),
          ),
          pw.SizedBox(width: 18),
          pw.Expanded(
            child: pw.Column(crossAxisAlignment: pw.CrossAxisAlignment.start, children: [
              infoRow('Tarih Aralığı', periodLabel),
              infoRow('Düzenleme Tarihi', issuedAt),
              infoRow('Para Birimi', 'TL'),
            ]),
          ),
        ]),
        pw.SizedBox(height: 12),

        // ---------- HAREKETLER ----------
        pw.Table(columnWidths: widths, children: tableRows),

        // ---------- YAZIYLA ----------
        pw.SizedBox(height: 14),
        pw.RichText(
          text: pw.TextSpan(children: [
            const pw.TextSpan(text: 'Yalnız ', style: pw.TextStyle(fontSize: 8.5, color: _muted)),
            pw.TextSpan(
              text: turkishAmountInWords(doc.closing.abs()),
              style: pw.TextStyle(fontSize: 9.5, fontWeight: pw.FontWeight.bold, color: _ink),
            ),
            if (!closingDebt)
              const pw.TextSpan(
                  text: ' (müşteri alacaklı)', style: pw.TextStyle(fontSize: 8.5, color: _muted)),
          ]),
        ),
        pw.SizedBox(height: 6),
        pw.Text('Not: Bu belge bilgilendirme amaçlıdır.',
            style: const pw.TextStyle(fontSize: 8, color: _muted)),
      ],
    ));

    await Printing.sharePdf(
      bytes: await pdfDoc.save(),
      filename: 'Cari-Hesap-Ekstresi-$customerCode.pdf',
    );
  }
}

File: Frontend/lib/statementPdf.ts
import type { Content, Margins, StyleDictionary, TableCell, TDocumentDefinitions } from 'pdfmake/interfaces'
import { createPdf } from '@/lib/pdfMake'
import { formatAmount, formatDocDate, turkishAmountInWords, type StatementRow } from '@/lib/accountStatement'

/**
 * CARİ HESAP EKSTRESİ — yazdırılabilir belge.
 *
 * Ekrandaki belge ile BİREBİR aynı düzen: kurum başlığı, cari bilgi ızgarası, Tarih / İşlem Türü /
 * Açıklama / Borç / Alacak / Bakiye tablosu, toplam + bakiye bandı, tutarın yazıyla okunuşu.
 * Rakamlar `buildAccountStatement`ten hazır gelir — bu dosya HESAP YAPMAZ, yalnız dizer
 * (iki yerde hesap yapılsaydı ekran ile kâğıt ayrı rakam yazabilirdi).
 */

const mg = (top: number, right: number, bottom: number, left: number): Margins => [top, right, bottom, left]

const COLORS = {
  ink: '#241C21',
  soft: '#5D4C55',
  muted: '#7A6873',
  line: '#DED5DA',
  band: '#F4EFF1',
  zebra: '#FBF8F9',
  plum: '#A5556E',
  debt: '#9F1239',
  credit: '#15694A',
}

export interface StatementPdfInstitution {
  name: string
  phone?: string | null
  email?: string | null
  taxNumber?: string | null
  taxOffice?: string | null
  branch?: string | null
}

export interface StatementPdfCustomer {
  code: string
  name: string
  phone?: string | null
  saleCount: number
}

export interface StatementPdfData {
  institution: StatementPdfInstitution
  customer: StatementPdfCustomer
  /** "18.06.2026 - 01.04.2027" */
  periodLabel: string
  /** "11.08.2026 17:30" */
  issuedAt: string
  rows: StatementRow[]
  totalDebit: number
  totalCredit: number
  closing: number
}

/** Belgenin sol/sağ bilgi bloğu: "Etiket : Değer" satırları. */
function infoColumn(entries: [string, string][]): Content {
  return {
    table: {
      widths: [72, 6, '*'],
      body: entries.map(([label, value]) => ([
        { text: label, style: 'infoLabel' },
        { text: ':', style: 'infoLabel' },
        { text: value || '—', style: 'infoValue' },
      ])),
    },
    layout: 'noBorders',
  }
}

export function generateAccountStatementPdf(
  data: StatementPdfData,
  action: 'download' | 'print' | 'open' = 'download',
): void {
  const { institution, customer } = data

  const head: TableCell[] = [
    { text: 'Tarih', style: 'th' },
    { text: 'İşlem Türü', style: 'th' },
    { text: 'Açıklama', style: 'th' },
    { text: 'Borç (TL)', style: 'th', alignment: 'right' },
    { text: 'Alacak (TL)', style: 'th', alignment: 'right' },
    { text: 'Bakiye (TL)', style: 'th', alignment: 'right' },
  ]

  const body: TableCell[][] = [head]

  if (data.rows.length === 0) {
    body.push([
      {
        text: 'Bu dönemde hareket bulunmuyor.',
        colSpan: 6, alignment: 'center', color: COLORS.muted, margin: mg(0, 10, 0, 10),
      },
      {}, {}, {}, {}, {},
    ])
  }

  data.rows.forEach((row, index) => {
    const fill = index % 2 === 1 ? COLORS.zebra : undefined
    body.push([
      { text: formatDocDate(row.date), style: 'td', fillColor: fill },
      { text: row.type, style: 'td', fillColor: fill },
      { text: row.description, style: 'td', fillColor: fill },
      { text: formatAmount(row.debit), style: 'tdNum', alignment: 'right', fillColor: fill },
      { text: formatAmount(row.credit), style: 'tdNum', alignment: 'right', fillColor: fill },
      { text: formatAmount(row.balance), style: 'tdNumBold', alignment: 'right', fillColor: fill },
    ])
  })

  // TOPLAM: yalnız borç/alacak sütunları toplanır — bakiye zaten son satırda yazılıdır.
  body.push([
    { text: '', border: [false, true, false, false], borderColor: [COLORS.line, COLORS.line, COLORS.line, COLORS.line] },
    { text: '' },
    { text: 'Toplam', style: 'totalLabel', alignment: 'right' },
    { text: formatAmount(data.totalDebit), style: 'totalValue', alignment: 'right' },
    { text: formatAmount(data.totalCredit), style: 'totalValue', alignment: 'right' },
    { text: '' },
  ])

  const closingDebt = data.closing >= 0
  body.push([
    { text: '' },
    { text: '' },
    { text: '' },
    { text: '' },
    { text: 'Bakiye', style: 'totalLabel', alignment: 'right' },
    {
      text: `${formatAmount(Math.abs(data.closing))} TL`,
      style: 'balanceValue',
      alignment: 'right',
      color: closingDebt ? COLORS.debt : COLORS.credit,
    },
  ])

  const content: Content[] = [
    // ---------- KURUM BAŞLIĞI ----------
    {
      columns: [
        {
          width: '*',
          stack: [
            { text: institution.name || 'Kurum', style: 'brand' },
            ...(institution.branch ? [{ text: institution.branch, style: 'brandSub' }] : []),
          ],
        },
        {
          width: 'auto',
          stack: [
            { text: (institution.name || 'Kurum').toLocaleUpperCase('tr'), style: 'orgName', alignment: 'right' },
            ...(institution.phone ? [{ text: `Tel: ${institution.phone}`, style: 'orgLine', alignment: 'right' as const }] : []),
            ...(institution.email ? [{ text: institution.email, style: 'orgLine', alignment: 'right' as const }] : []),
            ...(institution.taxNumber
              ? [{ text: `VKN: ${institution.taxNumber}${institution.taxOffice ? ` · ${institution.taxOffice}` : ''}`, style: 'orgLine', alignment: 'right' as const }]
              : []),
          ],
        },
      ],
      margin: mg(0, 0, 0, 8),
    },
    { canvas: [{ type: 'line', x1: 0, y1: 0, x2: 515, y2: 0, lineWidth: 1.6, lineColor: COLORS.plum }] },

    // ---------- BAŞLIK ----------
    { text: 'CARİ HESAP EKSTRESİ', style: 'docTitle', alignment: 'center', margin: mg(0, 16, 0, 14) },

    // ---------- CARİ BİLGİLERİ ----------
    {
      columns: [
        infoColumn([
          ['Cari Kodu', customer.code],
          ['Adı Soyadı', customer.name],
          ['Telefon', customer.phone || '—'],
          ['Kayıtlı Satış', `${customer.saleCount} satış`],
        ]),
        infoColumn([
          ['Tarih Aralığı', data.periodLabel],
          ['Düzenleme Tarihi', data.issuedAt],
          ['Para Birimi', 'TL'],
        ]),
      ],
      columnGap: 18,
      margin: mg(0, 0, 0, 14),
    },

    // ---------- HAREKETLER ----------
    {
      table: { headerRows: 1, widths: [52, 62, '*', 62, 62, 68], body },
      layout: {
        hLineWidth: (i: number) => (i === 0 || i === 1 || i === body.length ? 0.8 : 0.4),
        vLineWidth: () => 0,
        hLineColor: (i: number) => (i === 1 ? COLORS.plum : COLORS.line),
        paddingTop: () => 5,
        paddingBottom: () => 5,
        paddingLeft: () => 6,
        paddingRight: () => 6,
      },
    },

    // ---------- YAZIYLA ----------
    {
      text: [
        { text: 'Yalnız ', style: 'wordsLabel' },
        { text: turkishAmountInWords(Math.abs(data.closing)) || '—', style: 'wordsValue' },
        { text: closingDebt ? '' : ' (müşteri alacaklı)', style: 'wordsLabel' },
      ],
      margin: mg(0, 16, 0, 0),
    },
    { text: 'Not: Bu belge bilgilendirme amaçlıdır.', style: 'note', margin: mg(0, 8, 0, 0) },
  ]

  const styles: StyleDictionary = {
    brand: { fontSize: 15, bold: true, color: COLORS.ink },
    brandSub: { fontSize: 8.5, color: COLORS.muted, margin: mg(0, 2, 0, 0) },
    orgName: { fontSize: 10, bold: true, color: COLORS.ink },
    orgLine: { fontSize: 8, color: COLORS.muted, margin: mg(0, 1.5, 0, 0) },
    docTitle: { fontSize: 15, bold: true, color: COLORS.ink, characterSpacing: 0.6 },
    infoLabel: { fontSize: 8.5, color: COLORS.muted, margin: mg(0, 2, 0, 2) },
    infoValue: { fontSize: 9, color: COLORS.ink, bold: true, margin: mg(0, 2, 0, 2) },
    th: { fontSize: 8.5, bold: true, color: COLORS.ink, fillColor: COLORS.band },
    td: { fontSize: 8.5, color: COLORS.soft },
    tdNum: { fontSize: 8.5, color: COLORS.ink },
    tdNumBold: { fontSize: 8.5, bold: true, color: COLORS.ink },
    totalLabel: { fontSize: 9, bold: true, color: COLORS.ink },
    totalValue: { fontSize: 9, bold: true, color: COLORS.ink },
    balanceValue: { fontSize: 11, bold: true },
    wordsLabel: { fontSize: 8.5, color: COLORS.muted },
    wordsValue: { fontSize: 9.5, bold: true, color: COLORS.ink },
    note: { fontSize: 8, color: COLORS.muted },
  }

  const definition: TDocumentDefinitions = {
    info: {
      title: `Cari Hesap Ekstresi — ${customer.name}`,
      author: institution.name || 'BeautyAsist',
      subject: `Cari hesap ekstresi (${data.periodLabel})`,
      creator: 'BeautyAsist',
    },
    pageSize: 'A4',
    pageMargins: [40, 40, 40, 46],
    content,
    styles,
    defaultStyle: { font: 'Roboto', fontSize: 9, color: COLORS.ink },
    footer: (currentPage: number, pageCount: number) => ({
      columns: [
        {
          text: `${institution.name || 'BeautyAsist'} • Cari Hesap Ekstresi • ${customer.code}`,
          fontSize: 7.5, color: COLORS.muted, margin: mg(40, 0, 0, 0),
        },
        {
          text: `Sayfa ${currentPage} / ${pageCount}`,
          fontSize: 7.5, color: COLORS.muted, alignment: 'right', margin: mg(0, 0, 40, 0),
        },
      ],
      margin: mg(0, 14, 0, 0),
    }),
  }

  const doc = createPdf(definition)
  if (action === 'print') doc.print()
  else if (action === 'open') doc.open()
  else {
    const stamp = new Date()
    const day = `${stamp.getFullYear()}-${String(stamp.getMonth() + 1).padStart(2, '0')}-${String(stamp.getDate()).padStart(2, '0')}`
    doc.download(`Cari-Hesap-Ekstresi-${customer.code}-${day}.pdf`)
  }
}
